index.test-d.ts

```diff
@@ -0,0 +1,154 @@
+expectType<(a1:A1, a2:A2) => Promise<[Error | undefined]>>(pify(function20, {errorFirst: false, multiArgs: true}));
+
+
+expectType<typeof function00>(pify(function00, {excludeMain: true}));
+expectType<typeof function10>(pify(function10, {excludeMain: true}));
+expectType<typeof function20>(pify(function20, {excludeMain: true}));
+expectType<typeof function01>(pify(function01, {excludeMain: true}));
+expectType<typeof function11>(pify(function11, {excludeMain: true}));
+expectType<typeof function21>(pify(function21, {excludeMain: true}));
+expectType<typeof function02>(pify(function02, {excludeMain: true}));
+expectType<typeof function12>(pify(function12, {excludeMain: true}));
+expectType<typeof function22>(pify(function22, {excludeMain: true}));
+
+
+expectAssignable<Function>(pify(function00, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function01, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function02, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function10, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function11, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function12, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function20, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function21, {excludeMain: someBoolean}));
+expectAssignable<Function>(pify(function22, {excludeMain: someBoolean}));
+
+expectType<{
+	function: () => Promise<void>,
+	number: number,
+}>(pify({
+	function: function00,
+	number: 123,
+}));
+
+expectAssignable<{
+	function: () => Promise<void>,
+	number: number,
+	string?: string,
+}>(pify({
+	function: function00,
+	number: 1,
+}));
+
+expectNotType<{
+	function: () => Promise<void>,
+	string: string,
+	number: number,
+}>(pify({
+	function: function00,
+	string: 'qwe',
+}));
+
+declare function generic10<A0>(a0:A0, cb:(error: Error | null | undefined) => any): any;
+declare function generic01<R0>(cb:(error:  Error | null | undefined, r0:R0) => any): any;
+declare function generic11<A0, R0>(a0:A0, cb:(error:  Error | null | undefined, r0:R0) => any): any;
+
+expectType<Promise<A1>>((pify(generic01))<A1>());
+expectType<Promise<void>>((pify(generic10))<A1>(a1));
+expectType<Promise<A2>>((pify(generic11))<A1, A2>(a1));
+
+declare function realLife1(url: string, callback: (error?: Error | null | undefined, response?: { statusCode: number }, body?: string) => void): void;
+expectType<(url: string) => Promise<[{ statusCode: number }?, string?]>>(pify(realLife1, { multiArgs: true }))
+declare function realLife2(url: string, callback: (response?: { statusCode: number }, body?: string) => void): void;
+expectType<(url: string) => Promise<[{ statusCode: number }?, string?]>>(pify(realLife2, { multiArgs: true, errorFirst: false }))
+
+const fs = {
+    readFile(path: string, callback: (error?: Error, data?: Buffer) => void): void {
+        callback(undefined, Buffer.from('abc', 'utf8'));
```

Comment:
Tab-indentation.

Reply:
I will review the style and do some cleanup when we agree on functionality.